g1ant-language/desktop-apps-automation2.md

```diff
@@ -0,0 +1,211 @@
+
+![Sample mailing list CSV file in Notepad](../-assets/mailing-list.jpg)
+
+And here the G1ANT.Robot steps in and takes this job from you!
+
+Start with designing an architecture for your script: the plan of actions.
+
+1. Login into your email account.
+2. Check for unread emails containing “unsubscribe” in the subject field.
+3. If there are any, store the sender’s email address.
+4. Open the mailing list .csv file.
+5. Search for the stored email addresses and delete them from the list.
+6. Save and close the file.
+
+Seems pretty easy. And, in fact, it is with our G1ANT.Language commands.
+
+### Silent Execution
+
+You could do all of the above actions using UI commands you learned in the [Basics](../g1ant-language/basics.md) chapter: `program`, `window` and `keyboard`. They would open your email client (or webmail in a browser), search for “unsubscribe” message subject, copy the sender’s email, then open the mailing list file in Notepad and use its Find and Replace tool to delete the address from the list. But there are less obtrusive ways that do not involve dealing with the user interface and clicking around menus, dialogs and so on. They work directly with the target objects in the background, so users can still do tasks on their PCs.
+
+Take emails, for example. You don’t have to use any external tool such as email client or net browser to check your inbox: there is the `mail.imap` command for it. You simply provide your email account credentials as the arguments and the command gets messages for you, storing them in a variable. Then, you can search this variable for some text and even replace it with the `text.` family of commands.  And this is exactly what you are going to do in this exercise.
+
+### Reading Emails
+
+The whole process described in your plan of actions can be divided into three main tasks: retrieving emails, searching for specific messages and deleting email addresses from the list. Let’s make procedures out of these tasks and name them `checkmail`, `checksubject` and `deleteaddress`.
+
+```G1ANT
+procedure checkmail
+    
+end procedure
+
+procedure checksubject
+    
+end procedure
+
+procedure deleteaddress
+    
+end procedure
+```
+
+Start with the `checkmail` procedure. It should simply login to your email account and retrieve messages. The `mail.imap` command does it in one fell swoop. What it needs is the email server information, your login credentials and a name of the variable, which will store emails. You can also specify some filtering, which is helpful when processing huge mailboxes: time range (start and end dates of messages), an option to process only unread messages and whether to mark them as read afterwards. The full syntax for the command is as follows:
+
+```G1ANT
+mail.imap host ‴imap.server.name‴ port ‴993‴ login ‴your@email.address‴ password ‴p@$$w0rd‴ sincedate ♥startdate todate ♥enddate onlyunreadmessages true markallmessagesasread true result ♥result
```

Comment:
read again:
**The full syntax for the command is as follows:**
Do you understand the concept of SYNTAX example?
‴‴ are used in the command description -- this is how the developer described the command syntax

Reply:
Ok. But why ♥startdate and ♥enddate are here if these arguments take values of date structure and variable structure is not necessary here?
if we want to be consistent with this way of presenting our syntax, it should be:
`mail.imap host ‴‴ port 993 login ‴‴ password ‴‴ sincedate ‴‴ todate ‴‴ onlyunreadmessages false markallmessagesasread true result ♥result`

and there was a small mistake because onlyunreadmessages argument is false by default :)

Reply:
Remember: you write a manual for dumb people like me. Not for programmers! The manual should be comprehensive and EXPLAIN THINGS. What information does your example provide? Emptiness. No values. A reader doesn't have a clue what you mean by ‴‴ in arguments. You must show what data can be inserted and how they should look like.
In two things you're right: ♥startdate and ♥enddate should be preceded with [[date]] (or there should be only [[date]] with no variables) and port number shouldn't be in ‴‴.

Reply:
But it doesn't make it easier to understand because if someone wants to use this command, they have to distinguish default values from example ones. That's why we use syntax and then some examples od using some command. But ok leave it as you like (don't want to worry you but 200 commands descriptions syntax is a bit different so it will have to be changed as well)

Reply:
syntax in command descriptions will be written with the strict guidelines. Here, in exercises, we can use a more user friendly form of syntax.

Reply:
```suggestion
mail.imap host ‴imap.server.name‴ port 993 login ‴your@email.address‴ password ‴p@$$w0rd‴ sincedate ⟦date⟧ todate ⟦date⟧ onlyunreadmessages true markallmessagesasread true result ♥result
```